### PR TITLE
Introduce capacity parameter for topk

### PR DESCRIPTION
**Pull request by mkleen**
## Summary of the changes / Why this improves CrateDB

This changes the topk max limit to 5000 with capacity 8192 so the limit is less than 0.75 % of the capacity to have a better default behaviour for performance reasons. It also make the capacity configurable for the user.

### Background:
I tried out the nyc taxi datset with topk and found out that there is potential to improve performance significant by increasing the capacity of number of tracked items. Initializing an item is expensive while increasing a counter is cheap on the sketch datastructure. Small capacity means a small memory footprint, but If the capacity is too small items get dropped out and recreated which is slow because it requires multiple hashings. We should increase the default capacity size and make the parameter configurable for users like Spark does. On the nyc-taxi dataset this variant is much faster, and the other benchmarks are not slower.

```
V1: 5.9.0-9c2b7637a90c1271858560e0ce4288c569e7ade5
V2: 5.9.0-00be8919e533067034a1733d1c799101e57ce993

Q: select topk(pulocationid) from nyc_taxi
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       84.861 ±   21.829 |     81.305 |     82.471 |     82.880 |    300.588 |
|   V2    |       42.820 ±   20.812 |     26.345 |     40.736 |     41.025 |    247.978 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -  65.85%                           -  67.75%
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 65.85%
The test has statistical significance


System/JVM Metrics (durations in ms, byte-values in MB)
    |    YOUNG GC            |       OLD GC           |      HEAP         |     ALLOC
    |  cnt      avg      max |  cnt      avg      max |  initial     used |     rate      total
 V1 |    2     3.04     3.14 |    0     0.00     0.00 |      268     2001 |  3049.95      23119
 V2 |    0     0.00     0.00 |    0     0.00     0.00 |      268        0 |  5657.67      17064
```

```
Top frames (by count)
  V1
    ReversePurgeLongHashMap.purge(int) total=4601564458, count=2218
    ReversePurgeLongHashMap.adjustOrPutValue(long, long) total=944, count=944
    DocValuesAggregates.getRow(...) total=922, count=922
    ReversePurgeLongHashMap.keepOnlyPositiveCounts() total=262, count=262
    QuickSelect.partition(...) total=158, count=158
    AbstractMemorySegmentImpl.checkBounds(long, long) total=110, count=110
    ReversePurgeLongHashMap.hashDelete(int) total=90, count=90
    QuickSelect.select(...) total=49, count=49
    CompositeBatchIterator$AsyncCompositeBI.lambda$loadNextBatch$0(int, int) total=1906284128, count=18
    SingletonSortedNumericDocValues.nextValue() total=17, count=17
  V2
    DocValuesAggregates.getRow(...) total=546, count=546
    SingletonSortedNumericDocValues.nextValue() total=299, count=299
    ReversePurgeLongHashMap.adjustOrPutValue(long, long) total=237, count=237
    AbstractMemorySegmentImpl.checkBounds(long, long) total=129, count=129
    ParserATNSimulator.getEpsilonTarget(...) total=26216552, count=9
    ReversePurgeLongHashMap.purge(int) total=332351600, count=8
    ObjectName.construct(String) total=19064, count=8
    LinkedList.listIterator(int) total=14652848, count=6
    StreamSupport.stream(...) total=12304, count=6
    ObjectName.setCanonicalName(...) total=10344, count=5
```

For the implementation, there are a few options:

- Let the user use any value for capacity and convert internally to a power of number (@proddata favourite)
- Make it explicit, so capacity must be a value in the power of two (current approach)
- Use a load factor like clickhouse, this may be difficult to fine-tune for large datasets where you want a small topk.

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: docs/general/builtins/aggregation.rst [hunk 1]
```diff
@@ -832,15 +832,19 @@ Example::
 
 .. _aggregation-topk:
 
-``topk(column, k)``
+``topk(column, k, capacity)``
```
Comment (proddata) [inline docs/general/builtins/aggregation.rst:835]:
```suggestion
``topk(column, k, [capacity])``
```

File: docs/general/builtins/aggregation.rst [hunk 2]
```diff
@@ -832,15 +832,19 @@ Example::
...
 -------------------
 
 The ``topk`` aggregate function computes an approximation of the most frequent
 values and their frequencies of a specific column. The result is an array of
 objects containing the most common values and their frequencies ordered by
 frequency. The k parameter is an optional parameter which defines how many
 elements should be returned which defaults to 8. The maximum limit for k is
-10000.
+5000. The capacity parameter is optional and describes the maximum number
+of tracked items,must be in be a power of 2 and defaults to 8192.
```
[the remark this answers is not in the document]
Comment (mkleen) [reply docs/general/builtins/aggregation.rst:844]: That was in the previous implementation. If you look at a larger power of two number the value from the user to the actual can differ a lot which is confusing.

[the remark this answers is not in the document]
Comment (proddata) [reply docs/general/builtins/aggregation.rst:844]: I think the implications of changing the value from a user perspective aren't very clear anyway. However, needing to do a "calculation" to use a feature seems a bit odd. `hyperloglog_distinct`, as far as I know, has the `precision` option, which allows users to [specify the exponent](https://cratedb.com/docs/crate/reference/en/latest/general/builtins/aggregation.html#hyperloglog-distinct-column-precision).

> If you look at a larger power of two number, the value from the user to the actual can differ a lot, which is confusing.

Not fully following. The difference could be max the provided capacity (e.g., capacity=4097, diff=4095), but since the concrete implications are unknown, I think this isn't so important. A user might know how many distinct values a column has and should be able to provide that value. Then again, the question arises if we can't use column statistics to get a better estimate to start with. 🤷

[the remark this answers is not in the document]
Comment (mkleen) [reply docs/general/builtins/aggregation.rst:844]: 2^13 = 8192 and 2^14 = 16384. Now think if a user choose 10000. Since this is the number of tracked items which has a direct impact on the error rate and performance it think it's better to explicit.

[the remark this answers is not in the document]
Comment (mkleen) [reply docs/general/builtins/aggregation.rst:844]: Column statistics won't help, this is depending on what you want to do. Did you look at some top-k implementations in other systems ?

[the remark this answers is not in the document]
Comment (proddata) [reply docs/general/builtins/aggregation.rst:844]: > 2^13 = 8192 and 2^14 = 16384. Now think if a user choose 10000. Since this is the number of tracked items which has a direct impact on the error rate and performance it think it's better to explicit.

But if a user chooses 10000, this implies they need to use 16384 anyway?

> Did you look at some top-k implementations in other systems ?

ClickHouse is using a multiple of `k` 
https://clickhouse.com/docs/en/sql-reference/aggregate-functions/reference/topk
https://clickhouse.com/docs/en/sql-reference/aggregate-functions/reference/approxtopk

which defaults to `k * 3`

[the remark this answers is not in the document]
Comment (mkleen) [reply docs/general/builtins/aggregation.rst:844]: The previous implementation was similar and when I benchmarked more datasets I found it did not perform well on some data so i made the parameters explicit like in spark.

https://github.com/crate/crate/pull/16646
https://docs.databricks.com/en/sql/language-manual/functions/approx_top_k.html

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java [hunk 3]
```diff
@@ -117,7 +133,8 @@ public static void register(Functions.Builder builder) {
     private final DataType<?> argumentType;
 
     private static final int DEFAULT_LIMIT = 8;
-    private static final int MAX_LIMIT = 10_000;
+    private static final int MAX_LIMIT = 5_000;
+    private static final int DEFAULT_CAPACITY = 8192;
```
Comment (mkleen) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java:137]: Note: This the max capacity, the data sketch will start with 8 as default and then increase.

File: docs/general/builtins/aggregation.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (mfussenegger) [reply docs/general/builtins/aggregation.rst:844]: I'd also think that the capacity should grow in a somewhat reasonable way with the `k` parameter.
Otherwise you're kinda forced to provide it, no?

[the remark this answers is not in the document]
Comment (mkleen) [reply docs/general/builtins/aggregation.rst:844]: The main takeaway from the result here is that the initial capacity should be much larger by default.

[the remark this answers is not in the document]
Comment (mkleen) [reply docs/general/builtins/aggregation.rst:844]: Spark starts with limit 5 and capacity 10.000 as default value.

[the remark this answers is not in the document]
Comment (mkleen) [reply docs/general/builtins/aggregation.rst:844]: Now, the maximum for limit is 5000 with a capacity of 8192.

[the remark this answers is not in the document]
Comment (mfussenegger) [reply docs/general/builtins/aggregation.rst:844]: > The main takeaway from the result here is that the initial capacity should be much larger by default.

Isn't that trading memory for performance? Running a benchmark where you don't look at the memory consumption and get lower response times doesn't necessarily indicate that it is better to use a higher capacity by default.

And your replies are not really answering why it's not necessary or desirable to couple the default capacity to the value of `k`.

[the remark this answers is not in the document]
Comment (mkleen) [reply docs/general/builtins/aggregation.rst:844]: I think the main point of topk with sketch is that it has fixed memory assumption based on the capacity. If data becomes really large, the user can specify by the capacity how much memory he would like to use.

[the remark this answers is not in the document]
Comment (mkleen) [reply docs/general/builtins/aggregation.rst:844]: So, since the memory consumption can be fixed, this is the argument against a dynamic growing capacity.

[the remark this answers is not in the document]
Comment (mkleen) [reply docs/general/builtins/aggregation.rst:844]: Regarding the performance, it caught originally my attention based on the profiling and then found this curve https://datasketches.apache.org/docs/Frequency/FrequentItemsPerformance.html

[the remark this answers is not in the document]
Comment (mkleen) [reply docs/general/builtins/aggregation.rst:844]: I think if we make a dynamically generated capacity, it would be better to calculate that from the table size to go with the optimum.

[the remark this answers is not in the document]
Comment (mkleen) [reply docs/general/builtins/aggregation.rst:844]: > And your replies are not really answering why it's not necessary or desirable to couple the default capacity to the > value of k.

Afaik `k` is not the leading factor here. The main factor is number of processed items and capacity (=available memory) for each job.

[the remark this answers is not in the document]
Comment (mkleen) [reply docs/general/builtins/aggregation.rst:844]: Anyway, what would be your suggestion here ? Leave it as it ? Just make it possible to add the capacity as a limit without the new defaults ?

[the remark this answers is not in the document]
Comment (mfussenegger) [reply docs/general/builtins/aggregation.rst:844]: A query like:

```sql
select "cCode", topk("searchWord", 10) from uservisits group by 1 order by 1
```

With the uservisits dataset we have in crate-benchmarks and 2G heap will run
into a circuit breaker exception with the change here. Without the change it
works.

On the other hand, with the current default the accuracy looks rather bad

This is the real count:

```
cr> select "searchWord", count(*) from uservisits where "cCode" = 'AUT' group by 1 order by 2 desc limit 10;
+------------+----------+
| searchWord | count(*) |
+------------+----------+
| antiphoner |       53 |
| cathodes   |       53 |
| revoke     |       53 |
| unwaking   |       53 |
| Behistun   |       52 |
| neology    |       52 |
| dithymol   |       52 |
| ovaritides |       51 |
| pyronines  |       51 |
| palaeogene |       51 |
+------------+----------+
```

These are the approximates:

```
cr> select unnest(topk("searchWord", 8, 16)) from uservisits where "cCode" = 'AUT' ;
+--------------------------------------------------+
| unnest(topk("searchWord", 8, 16))                |
+--------------------------------------------------+
| {"frequency": 2713, "item": "piezometrical"}     |
| {"frequency": 2713, "item": "preignition"}       |
| {"frequency": 2713, "item": "sidebands"}         |
| {"frequency": 2713, "item": "tree-embowered"}    |
| {"frequency": 2713, "item": "VAP"}               |
| {"frequency": 2713, "item": "unwaking"}          |
| {"frequency": 2713, "item": "chronogrammatical"} |
+--------------------------------------------------+

cr> select unnest(topk("searchWord", 8, 2 ^ 8)) from uservisits where "cCode" = 'AUT' ;
+--------------------------------------------+
| unnest(topk("searchWord", 8, 256))         |
+--------------------------------------------+
| {"frequency": 185, "item": "inebriative"}  |
| {"frequency": 184, "item": "bradytely"}    |
| {"frequency": 184, "item": "ovaritides"}   |
| {"frequency": 184, "item": "Talegallinae"} |
| {"frequency": 184, "item": "matters"}      |
| {"frequency": 184, "item": "twice-worn"}   |
| {"frequency": 184, "item": "flypaper"}     |
| {"frequency": 184, "item": "refurnished"}  |
+--------------------------------------------+
SELECT 8 rows in set (0.006 sec)
cr> select unnest(topk("searchWord", 8, 2 ^ 9)) from uservisits where "cCode" = 'AUT' ;
+-------------------------------------------+
| unnest(topk("searchWord", 8, 512))        |
+-------------------------------------------+
| {"frequency": 96, "item": "dispending"}   |
| {"frequency": 96, "item": "ovaritides"}   |
| {"frequency": 96, "item": "metageometry"} |
| {"frequency": 95, "item": "flypaper"}     |
| {"frequency": 95, "item": "Teutolatry"}   |
| {"frequency": 95, "item": "carabaos"}     |
| {"frequency": 94, "item": "uther"}        |
| {"frequency": 94, "item": "meticulously"} |
+-------------------------------------------+
SELECT 8 rows in set (0.006 sec)
cr> select unnest(topk("searchWord", 8, 2 ^ 10)) from uservisits where "cCode" = 'AUT' ;
+-------------------------------------------+
| unnest(topk("searchWord", 8, 1024))       |
+-------------------------------------------+
| {"frequency": 65, "item": "metageometry"} |
| {"frequency": 65, "item": "Mantua"}       |
| {"frequency": 64, "item": "ovaritides"}   |
| {"frequency": 64, "item": "dithymol"}     |
| {"frequency": 64, "item": "oar-footed"}   |
| {"frequency": 63, "item": "antiphoner"}   |
| {"frequency": 63, "item": "dispending"}   |
| {"frequency": 63, "item": "darkens"}      |
+-------------------------------------------+
```


> I think if we make a dynamically generated capacity, it would be better to calculate that from the number of rows in the table.

Why? You might have intermediate operators, and number of rows != cardinality of the expression used for `topk`
Please state rationale and properties by which something is better if you make "would be better" statements.


---

So in short:

- Yes we should add a `capacity` parameter
- We should improve the docs to give the users some kind of info on what capacity yields what can of accuracy.

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java (hunk 3 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java:137]: Why not call it `DEFAULT_MAX_CAPACITY` then?
And probably also refer in the docs as `maxCapacity` if it is not pre-allocated fully

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java [hunk 4]
```diff
@@ -161,24 +178,24 @@ public State iterate(RamAccounting ramAccounting,
                          Input<?>... args) throws CircuitBreakingException {
         Object value = args[0].value();
         if (state instanceof Empty) {
+            if (args.length == 3) {
+                Integer limit = (Integer) args[1].value();
+                Integer capacity = (Integer) args[2].value();
+                state = initState(ramAccounting, limit, capacity);
+            }
             if (args.length == 2) {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java:186]:
```suggestion
            } else if (args.length == 2) {
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java [hunk 5]
```diff
@@ -220,28 +237,52 @@ public DocValueAggregator<?> getDocValueAggregator(LuceneReferenceResolver refer
...
+        if (optionalParams.size() == 3) {
+            int limit, capacity;
+            Literal<?> limitLiteral = optionalParams.get(1);
+            Literal<?> capacityLiteral = optionalParams.get(2);
+
+            if (limitLiteral != null) {
+                limit = (int) limitLiteral.value();
+            } else {
+                limit = DEFAULT_LIMIT;
+            }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java:263]: Why is there a default imit in the size() == 3 case, but not in the size == 2 case?

Comment (mkleen) [reply server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java:263]: 👍

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java (hunk 3 above)
Comment (mkleen) [reply server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java:137]: 👍

Comment (mkleen) [reply server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java:137]: So should I keep the new default value of 8192 ?

File: docs/general/builtins/aggregation.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (mkleen) [reply docs/general/builtins/aggregation.rst:844]: > Yes we should add a capacity parameter

Do you have an opinion if the capacity parameter should be in the power of two or do prefer accepting any value and do internally rounding ?

>  We should improve the docs to give the users some kind of info on what capacity yields what can of accuracy.

What do you think about returning the an error/acuracy value from the sketch as part of the result ?

[the remark this answers is not in the document]
Comment (mfussenegger) [reply docs/general/builtins/aggregation.rst:844]: > Do you have an opinion if the capacity parameter should be in the power of two or do prefer accepting any value and do internally rounding ?

Require power of two. Doing rounding would just obfuscate what's happening

> What do you think about returning the an error/acuracy value from the sketch as part of the result ?

That sounds nice. How would you represent that?

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java (hunk 3 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java:137]: > So should I keep the new default value of 8192 ?

Yes

Comment (mkleen) [reply server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java:137]: 👍

File: docs/general/builtins/aggregation.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (mkleen) [reply docs/general/builtins/aggregation.rst:844]: > Please state rationale and properties by which something is better if you make "would be better" statements.
The relation between number of processed items, number of stored items and n max size (capacity) is described in this https://datasketches.apache.org/docs/Frequency/FrequentItemsErrorTable.html For example with a max capacity of 8,192 it is possible to detect the frequency of 1 million items if it's 854 or less. The sketch itself provides relative maximum error which we could return as part of the result e.g.:

Exact:
```
cr> select "searchWord", count(*) from uservisits  group by 1 order by 2 desc limit 10;
+----------------+----------+
| searchWord     | count(*) |
+----------------+----------+
| CTTN           |     4098 |
| german         |     4086 |
| tarlatans      |     4044 |
| uredo-fruit    |     4041 |
| magnetizations |     4033 |
| Shirleen       |     4032 |
| Downe          |     4031 |
| Biamonte       |     4027 |
| etagere        |     4017 |
| verdugoship    |     4009 |
+----------------+----------+
```
Approximation with no error:
```
cr> select unnest(topk("searchWord", 8, 2048)) from uservisits;
+-----------------------------------------------+
| unnest(topk("searchWord", 8, 2048))           |
+-----------------------------------------------+
| {"maximum_error": 0}                          |
| {"frequency": 4098, "item": "CTTN"}           |
| {"frequency": 4086, "item": "german"}         |
| {"frequency": 4044, "item": "tarlatans"}      |
| {"frequency": 4041, "item": "uredo-fruit"}    |
| {"frequency": 4033, "item": "magnetizations"} |
| {"frequency": 4032, "item": "Shirleen"}       |
| {"frequency": 4031, "item": "Downe"}          |
| {"frequency": 4027, "item": "Biamonte"}       |
+-----------------------------------------------+

```
Approximation with error:
```
cr> select unnest(topk("searchWord", 8, 1024)) from uservisits;
+----------------------------------------------+
| unnest(topk("searchWord", 8, 1024))          |
+----------------------------------------------+
| {"maximum_error": 6328}                      |
| {"frequency": 6340, "item": "thievingly"}    |
| {"frequency": 6340, "item": "Culpeper"}      |
| {"frequency": 6337, "item": "subacutely"}    |
| {"frequency": 6336, "item": "subglobosity"}  |
| {"frequency": 6335, "item": "underthrust"}   |
| {"frequency": 6335, "item": "unsaved"}       |
| {"frequency": 6335, "item": "nervelessness"} |
| {"frequency": 6335, "item": "Treharne"}      |
+----------------------------------------------+
```
The maximum relative error describes how much the frequency of an item can be off at maximum. So my suggestion would to provide some kind of metric in the docs how to choose a reasonable capacity and return the error.

[the remark this answers is not in the document]
Comment (mfussenegger) [reply docs/general/builtins/aggregation.rst:844]: How about exposing all the info we have:

Turn the result from `list<map<...><` into an object with:

```
{
    "frequencies": [
        {
            "item": <expr>,
            "estimate": <item.estimate>,
            "lower_bound": <item.getlowerBound>,
            "upper_bound": <item.getupperbound>,
        },
        ....
    ],
    "maximum_error": <max_error>
}
```

**Comment by mkleen** [comment]
Screenshot 2024-10-02 at 16 42 50

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java (hunk 4 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java:186]: This doesn't look resolved? There's still no else if

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java [hunk 6]
```diff
@@ -220,28 +237,57 @@ public DocValueAggregator<?> getDocValueAggregator(LuceneReferenceResolver refer
...
+        }
+        // topk(ref, limit) -> aggregationReferences[ref, null] optionalParams[null, limit]
+        if (optionalParams.size() == 2) {
+            int limit;
+            Literal<?> limitLiteral = optionalParams.getLast();
+            if (limitLiteral != null) {
+                limit = (int) limitLiteral.value();
+            } else {
+                limit = DEFAULT_LIMIT;
+            }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java:255]:
```suggestion
            Literal<?> limitLiteral = optionalParams.getLast();
            int limit = limitLiteral == null ? DEFAULT_LIMIT : (int) limitLiteral.value();
```

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java [hunk 7]
```diff
@@ -220,28 +237,57 @@ public DocValueAggregator<?> getDocValueAggregator(LuceneReferenceResolver refer
...
+            if (capacityLiteral != null) {
+                capacity = (int) capacityLiteral.value();
+            } else {
+                capacity = DEFAULT_MAX_CAPACITY;
+            }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java:273]:
```suggestion
            Literal<?> limitLiteral = optionalParams.get(1);
            Literal<?> capacityLiteral = optionalParams.get(2);

            int limit = limitLiteral == null ? DEFAULT_LIMIT : (int) limitLiteral.value();
            int capacity = capacityLiteral == null ? DEFAULT_MAX_CAPACITY : (int) capacityLiteral.value();
```

File: docs/general/builtins/aggregation.rst [hunk 8]
```diff
@@ -832,31 +832,103 @@ Example::
...
-
-Internally the ``topk``
-aggregate function uses an implementation of the
-`Efficient Computation of Frequent and Top-k Elements in Data Streams`_
-algorithm.
+The ``topk`` aggregate function computes the most frequent values and their
+frequencies of a specific column. The result contains the top k most common
+values and their frequencies ordered by frequency. The k parameter is an
+optional parameter which defines how many elements should be returned which
+defaults to 8. The maximum limit for k is 5000. The max_capacity parameter is
```
Comment (mfussenegger) [inline docs/general/builtins/aggregation.rst:842]:
```suggestion

The ``topk`` aggregate function computes ``k`` most frequent values. The result
is an ``OBJECT`` in the following format:

    {
        "frequencies": [
            {
                "estimate": <estimated_frequency>,
                "item": "<value_of_column>",
                "lower_bound": <lower_bound>,
                "upper_bound": <upper_bound>"
            },
            ...
        ],
        "maximum_error": <max_error>
    }

The ``frequencies`` list is ordered by the estimated frequency, with the most
common items listed first.

``k`` defaults to 8 and can't exceed 5000. The ``max_capacity`` parameter is
```

**Review by mfussenegger — APPROVED** [review]

File: docs/general/builtins/aggregation.rst [hunk 9]
```diff
@@ -832,31 +832,103 @@ Example::
...
+     - 53406
+     - 26703
+     - 13351
+     - 6676
+
+
+The error threshold shows which ranges of frequencies can be tracked depending
+on the number of distinct items and max capacity. E.g. with the default max_capacity of
+8192 frequencies higher than 4 can be tracked on 10000 distinct items reliable.
+A threshold of 0 indicates that the frequencies are exact.
```
Comment (proddata) [inline docs/general/builtins/aggregation.rst:931]: looking at https://datasketches.apache.org/docs/Frequency/FrequentItemsErrorTable.html
This is not talking about distinct elements in the stream, but rather the total amount of elements processed in the stream

In the scenario where you have processed 10,000 items with a max capacity of 8192:
- 	The threshold  T  is about 4.27.
- 	All items with frequencies greater than 4 will be included.
- 	Some items with frequencies below the threshold 4 may also appear in the result.

Comment (mkleen) [reply docs/general/builtins/aggregation.rst]: You are right, thanks for the feedback
